<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [343]:
import pandas as pd
import psycopg2
import plotly.express as px

In [345]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [346]:
query_3_1 = f'''
select
    count(*)
from vacancies
'''

In [347]:
df_3_1 = pd.read_sql_query(query_3_1, connection)

cnt_vacancies = df_3_1['count'][0]
print(f'Всего вакансий: {cnt_vacancies}')

Всего вакансий: 49197


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/4281643191.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [348]:
query_3_2 = f'''
select
    count(*)
from employers
'''

In [349]:
df_3_2 = pd.read_sql_query(query_3_2, connection)

cnt_employers = df_3_2['count'][0]
print(f'Количество работодателей: {cnt_employers}')

Количество работодателей: 23501


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/308652486.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



3. Посчитате с помощью запроса количество регионов (таблица areas).

In [350]:
query_3_3 = f'''
select
    count(*)
from areas
'''

In [351]:
df_3_3 = pd.read_sql_query(query_3_3, connection)

cnt_areas = df_3_3['count'][0]
print(f'Количество регионов: {cnt_areas}')

Количество регионов: 1362


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/2098612830.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [352]:
query_3_4 = f'''
select
    count(*)
from industries
'''

In [353]:
df_3_4 = pd.read_sql_query(query_3_4, connection)

cnt_industries = df_3_4['count'][0]
print(f'Количество сфер деятельности: {cnt_industries}')

Количество сфер деятельности: 294


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/2819232353.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



***

Выполнив запросы, мы узнали количество вакансий, работодателей, регионов и сфер деятельности

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [354]:
query_4_1 = f'''
select
    a.name,
    count(*) cnt
from vacancies v
join areas a on a.id = v.area_id --Объединим таблицы по общему id
group by a.name
order by cnt desc --Сортируем по кол-ву вакансий в порядке убывания
limit 5
'''

In [355]:
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/1522353868.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [356]:
query_4_2 = f'''
select
    count(*)
from vacancies
where salary_from is not null
    or salary_to is not null
'''

In [357]:
df_4_2 = pd.read_sql_query(query_4_2, connection)

not_null_sal = df_4_2['count'][0]
print(f'Кол-во вакансий с зарплатой: {not_null_sal}')

Кол-во вакансий с зарплатой: 24073


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/2432422517.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [358]:
query_4_3 = f'''
select
    salary_from,
    salary_to
from vacancies
'''

In [359]:
df_4_3 = pd.read_sql_query(query_4_3, connection)

# Найдем средние значения для нижней и верхней границы зарплатной вилки
salary_from_mean = round(df_4_3['salary_from'].mean())
salary_to_mean = round(df_4_3['salary_to'].mean())

print(f'Среднее для нижней границы зарплатной вилки: {salary_from_mean}')
print(f'Среднее для верхней границы зарплатной вилки: {salary_to_mean}')

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/3701071455.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Среднее для нижней границы зарплатной вилки: 71065
Среднее для верхней границы зарплатной вилки: 110537


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [360]:
query_4_4 = f'''
select
    schedule,
    employment,
    count(*) cnt
from vacancies v
join employers e on e.id = v.employer_id --Объединим таблицы по общему id
group by schedule, employment
order by cnt desc --Отсортрируем кол-во вакансий в порядке убывания
'''

In [361]:
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/1636312543.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [362]:
query_4_5 = f'''
select
    experience,
    count(*) cnt
from vacancies v
group by experience
order by cnt --сортируем по возрастанию кол-ва вакансий
'''

In [363]:
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/802873672.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [364]:
# График кол-ва вакансий в разных городах
city_bar = px.bar(
    data_frame=df_4_1,
    x='name',
    y='cnt',
    title='Количество вакансий в разных городах',
    color='name',
    template='xgridoff'
)
city_bar.update_layout(xaxis_title='Город', yaxis_title='Количество вакансий', showlegend=False)
city_bar.show()

# График кол-ва вакансий для требуемого опыта
exp_bar = px.bar(
    data_frame=df_4_5,
    x='experience',
    y='cnt',
    color='experience',
    title='Количество вакансий для требуемуего опыта',
    template='xgridoff'
)
exp_bar.update_layout(xaxis_title='Опыт работы', yaxis_title='Количество вакансий', showlegend=False)
exp_bar.show()

![pic](/pic/city_bar.png)

![pic](/pic/exp_bar.png)

Наглядно видно, что по количеству вакансий лидирует город Москва, а самый частый требуемый опыт работы составляет от 1 года до 3 лет. Также, выполнив запросы, мы узнали, что самая популярная пара графика и трудоустройства - полный день/полная занятость. Благодаря анализу можно сделать вывод о нынешнем положении на рынке труда, о том, сколько компании готовы платить работникам на той или иной должности, предпочтительный уровень требуемого опыта и типа занятости. 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [365]:
query_5_1 = f'''
select
    e.name,
    count(*) cnt
from vacancies v
join employers e on e.id = v.employer_id --Объединим таблицы по общему id
group by e.name
order by cnt desc
limit 5
'''

In [366]:
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/556590240.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [367]:
query_5_2 = f'''
select
    a.name area,
    count(e.id) emp_cnt
from areas a
--Объединим таблицы друг с другом по общему id
left join vacancies v on v.area_id = a.id
left join employers e on e.area = a.id
where v.id is null
group by a.name
order by emp_cnt desc
limit 1 --Ограничим вывод
'''

In [368]:
df_5_2 = pd.read_sql_query(query_5_2, connection)
print('Больше всего работодателей без вакансий находятся в регионе', df_5_2['area'][0], '-',df_5_2['emp_cnt'][0])

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/3348995541.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Больше всего работодателей без вакансий находятся в регионе Россия - 410


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [369]:
query_5_3 = f'''
select
    e.name,
    count(distinct v.area_id) cnt
from employers e
left join vacancies v on e.id = v.employer_id --Объединим таблицы по общему id
group by e.name
order by cnt desc
limit 10
'''

In [370]:
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/3556584873.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt
0                      Яндекс  181
1                  Ростелеком  152
2                  Спецремонт  116
3      Поляков Денис Иванович   88
4                   ООО ЕФИН    71
5                  Совкомбанк   63
6                         МТС   55
7  ЭФКО, Управляющая компания   49
8                Почта России   48
9                        КРОН   48

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [371]:
query_5_4 = f'''
select
    count(*)
from employers e
left join employers_industries ei on ei.employer_id = e.id --Объединим таблицы по общему id
where ei.industry_id is null
'''

In [372]:
df_5_4 = pd.read_sql_query(query_5_4, connection)

null_ind = df_5_4['count'][0]
print(f'Кол-во работодателей, у которых не указана сфера дея-ти: {null_ind}')

Кол-во работодателей, у которых не указана сфера дея-ти: 8419


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/3332386460.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [373]:
query_5_5 = f'''
select
    e.name,
    count(i.name) cnt
from employers e
--Объединим таблицы друг с другом по общему id
left join employers_industries ei on ei.employer_id = e.id
join industries i on ei.industry_id = i.id
group by e.name
having count(i.name) = 4
order by e.name
offset 2 --Уберем первые две строки
limit 1 --Ограничим вывод
'''

In [374]:
df_5_5 = pd.read_sql_query(query_5_5, connection)

company_3rd = df_5_5['name'][0]
print(f'Компания с четырьмя сферами дея-ти: {company_3rd}')

Компания с четырьмя сферами дея-ти: 2ГИС


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/2657852809.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [375]:
query_5_6 = f'''
select
    name,
    count(ei.employer_id)
from industries i
join employers_industries ei on i.id = ei.industry_id 
group by name
having name = 'Разработка программного обеспечения'
'''

In [376]:
df_5_6 = pd.read_sql_query(query_5_6, connection)

result = df_5_6['count'][0]
print(f'Разработка программного обеспечения указана у {result} работодателей')

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/3126285072.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Разработка программного обеспечения указана у 3553 работодателей


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [377]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url)

page = BeautifulSoup(response.text, 'html.parser')
html = page.find_all('table')[0].find_all('tr')[1:]

city_list = []
for elem in html:
    for i, j in enumerate(elem):
        if i == 1:
            city_list.append(j.get_text())
            
# Преобразуем список в кортеж для SQL
city_tuple = tuple(city_list)
city_tuple

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [378]:
query_5_7 = f'''
(select
    a.name,
    count(v.id) as cnt
from vacancies v
join areas a on v.area_id = a.id
join employers e on v.employer_id = e.id
where e.name = 'Яндекс'
    and a.name in {city_tuple}
group by a.name
order by cnt)
union all
select
    'total',
    sum(cnt)
from (
    select
        count(v.id) as cnt
    from vacancies v
    join areas a on v.area_id = a.id
    join employers e on v.employer_id = e.id
    where e.name = 'Яндекс'
        and a.name in {city_tuple}
    group by a.name
) as subquery
'''

In [379]:
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/43835236.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            total  485.0

***

Исходя из запросов, можно сделать вывод, что самым "крупным" работодателем явялется Яндекс, опубликовав 1933 вакансии в 181 регионах.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [380]:
query_6_1 = f'''
select
    count(name)
from vacancies
where name ilike '%data%'
    or name ilike '%данн%'
'''

In [381]:
df_6_1 = pd.read_sql_query(query_6_1, connection)

data_cnt = df_6_1['count'][0]
print(f'Количество вакансий, которые имеют отношение к данным: {data_cnt}')

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/1251674964.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий, которые имеют отношение к данным: 1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [382]:
query_6_2 = f'''
select
    count(*)
from vacancies
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or name ilike '%machine learning%'
    or name like '%ML%' and name not ilike '%html%'
    or name ilike '%машинн%обучен%')
    and (name ilike '%junior%'
    or experience = 'Нет опыта'
    or employment = 'Стажировка')
'''

In [383]:
df_6_2 = pd.read_sql_query(query_6_2, connection)

junior_cnt = df_6_2['count'][0]
print(f'Количество вакансий для начинающего дата-сайентиста составляет: {junior_cnt}')

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/303321493.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество вакансий для начинающего дата-сайентиста составляет: 51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [384]:
query_6_3 = f'''
select
    count(*)
from vacancies
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or name ilike '%machine learning%'
    or name like '%ML%' and name not ilike '%html%'
    or name ilike '%машинн%обучен%')
    and (key_skills ilike '%SQL%'
    or key_skills ilike '%postgres%')
'''

In [385]:
df_6_3 = pd.read_sql_query(query_6_3, connection)

key_skills_vacantions = df_6_3['count'][0]
print(f'Количество подходящих вакансий составляет: {key_skills_vacantions}')

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/816136067.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Количество подходящих вакансий составляет: 201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [386]:
query_6_4 = f'''
select
    count(*)
from vacancies
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or name ilike '%machine learning%'
    or name like '%ML%' and name not ilike '%html%'
    or name ilike '%машинн%обучен%')
    and key_skills ilike '%python%'
'''

In [387]:
df_6_4 = pd.read_sql_query(query_6_4, connection)

py_vacancies = df_6_4['count'][0]
print(f'Количество вакансий с навыком Python составляет: {py_vacancies}')

Количество вакансий с навыком Python составляет: 351


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/733170710.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [388]:
query_6_5 = f'''
select
    --Найдем среднее кол-во навыков и округлим до 2-х знаков
    round(avg(length(key_skills) - length(replace(key_skills, CHR(9), '')) + 1), 2) avg_skills
from vacancies
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or name ilike '%machine learning%'
    or name like '%ML%' and name not ilike '%html%'
    or name ilike '%машинн%обучен%')
    and key_skills is not null
'''

In [389]:
df_6_5 = pd.read_sql_query(query_6_5, connection)

avg_skills = df_6_5['avg_skills'][0]
print(f'В среднем в вакансиях указывают {avg_skills} навыков')

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/2516728354.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



В среднем в вакансиях указывают 6.41 навыков


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [390]:
query_6_6 = f'''
select
    experience,
    round(avg(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)))
from vacancies
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or name ilike '%machine learning%'
    or name like '%ML%' and name not ilike '%html%'
    or name ilike '%машинн%обучен%')
    and (salary_from is not null
    or salary_to is not null)
group by experience
'''

In [391]:
df_6_6 = pd.read_sql_query(query_6_6, connection)

exp_avg_salary = df_6_6['round'][2]
print(f'Средняя зарплата дата-сайентиста с опытом от 3 до 6 лет составляет: {exp_avg_salary}')

Средняя зарплата дата-сайентиста с опытом от 3 до 6 лет составляет: 243115.0


/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/841710414.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



***

Проведя анализ вакансий, связанных с Data Science, была выявлена зависимость между средней заработной платой и опытом работы, сколько всего вакансий есть для DS и какие навыки требуют работодатели.

# Общий вывод по проекту

Проанализировав вакансии из HeadHunter с помощью Python и SQL, были сделаны следующие ключевые выводы:

1. ***Общее количество вакансий, работодателей, регионов и сфер деятельности:*** В результате выполнения запросов была собрана информация о числе вакансий, работодателей, регионов и сфер деятельности, что позволило получить общую картину предложения на рынке труда.

2. ***Региональная активность и требования к опыту работы:*** Москва является лидером по количеству вакансий, что свидетельствует о высоком уровне экономической активности в столице. Наиболее частое требование к опыту работы – от 1 года до 3 лет.

3. ***Крупнейшие работодатели:*** Анализ показал, что Яндекс является самым крупным работодателем, разместив 1933 вакансии в 181 регионе. Это указывает на широкую географическую представленность и значительные потребности компании в кадрах.

4. ***Анализ вакансий Data Science:*** Вакансии, связанные с Data Science, показали зависимость между средней заработной платой и требуемым опытом работы. Также было выявлено количество доступных вакансий в данной сфере и определены основные навыки, требуемые работодателями, что позволяет соискателям лучше ориентироваться в требованиях рынка.

Так же отдельно можно рассмотреть распределение вакансий Data Science по отдельным регионам

In [392]:
# Сделаем запрос, который покажет кол-во вакансий DS в разных регионах
query_7_1 = f'''
select
    a.name,
    count(*) cnt
from vacancies v
join areas a on a.id = v.area_id --Объединим таблицы по общему id
where (v.name ilike '%data scientist%'
    or v.name ilike '%data science%'
    or v.name ilike '%исследователь данных%'
    or v.name ilike '%machine learning%'
    or v.name like '%ML%' and v.name not ilike '%html%'
    or v.name ilike '%машинн%обучен%')
group by a.name
order by cnt desc --Отсортируем по убыванию
limit 9 --Ограничим вывод
'''

In [393]:
df_7_1 = pd.read_sql_query(query_7_1, connection)

# График кол-ва вакансий DS по регионам
ds_cities = px.bar(
    data_frame=df_7_1,
    x='name',
    y='cnt',
    color='name',
    template='xgridoff',
    title='Количество вакансий DS по регионам'
)
ds_cities.update_layout(xaxis_title='Регион', yaxis_title='Количество', showlegend=False)
ds_cities.show()

/var/folders/02/gp3q6k2d0rz6jktg0qk2yyzh0000gn/T/ipykernel_28827/1084996234.py:1: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



![pic](/pic/city_ds_bar.png)

Распределение вакансий DS по регионам, показывает, что большинство из них размещены в Москве. В первую очередь это связано с населением, но из-за того что эта сфера относительно новая, в последствии она охватит и остальные регионы

В целом, проведенный анализ позволил выявить основные тенденции и текущую ситуацию на рынке труда, что может быть полезным как для соискателей, так и для работодателей в целях оптимизации своих стратегий поиска работы и подбора персонала.